gflownet/utils/batch.py

```diff
@@ -0,0 +1,288 @@
+            self.actions = tfloat(
+                self.actions, device=self.device, float_type=self.float
+            )
+            self.done = tbool(self.done, device=self.device)
+            self.masks_invalid_actions_forward = tbool(
+                self.masks_invalid_actions_forward, device=self.device
+            )
+            if self.loss == "flowmatch":
+                self.parents_state_idx = tlong(
+                    sum([[idx] * len(p) for idx, p in enumerate(self.parents)], []),
+                    device=self.device,
+                )
+                self.parents_actions = torch.cat(
+                    [
+                        tfloat(x, device=self.device, float_type=self.float)
+                        for x in self.parents_actions
+                    ]
+                )
+            elif self.loss == "trajectorybalance":
+                self.masks_invalid_actions_backward = tbool(
+                    self.masks_invalid_actions_backward, device=self.device
+                )
+            self._process_parents()
+        self.is_processed = True
+
+    def _process_states(self):
+        """
+        Converts self.statees from list to torch tensor, computes states in the policy format
+        (stored in self.states_policy)
+        """
+        self.states_policy = self.states2policy()
+        self.states = tfloat(self.states, device=self.device, float_type=self.float)
+
+    def states2policy(self, states=None, env_ids=None):
+        """
+        Converts states from a list of states in gflownet fromat to a tensor of states in policy format
+        states: list of gflownet states,
+        env_ids: list of env ids indicating which env corresponds to each state in states list
+
+        Returns: torch tensor of stattes in policy format
+        """
+        if states is None:
+            states = self.states
+            env_ids = self.env_ids
+        elif env_ids is None:
+            # if states are provided, env_ids should be provided too
+            raise Exception(
+                "env_ids must be provided to the batch for converting provided states to the policy format"
+            )
+        states_policy = []
+        for state, env_id in zip(states, env_ids):
+            states_policy.append(self.envs[env_id].state2policy(state))
+        return tfloat(states_policy, device=self.device, float_type=self.float)
+
+    def _process_parents(self):
+        """
+        Prepares self.parents (gflownet format) and self.parents_policy (policy format) as torch tensors.
+        Different behaviour depending on self.loss:
+            - for flowmatch, parents contain all the possible parents for each state,
+            so this tensor is bigger than self.state (all the parents are alinged the zero dimension, )
+            - for trajectorybalance, parents contain only one parent for each state which was its parent in the trajectory
+        """
+        if self.loss == "flowmatch":
```

Comment:
Would it make sense for the batch object not to know the loss and instead always collect everything and the GFlowNet will decide what to use? 
For instance, instead of having self.parents mean something different depending on the loss, have something like self.possible_parents for the possible parents of each state (as in FM) and self.parents for the actual parents of each state (as in TB).

I think that it might help with making the code a bit simpler and easier to maintain : 
- The batch doesn't need to know the loss (less coupling and less if/else)
- Easier to extend to new losses in the future since any such change will only need to be implemented in gflownet.py
- No attributes (like self.parents) that completely change semantics depending on the loss so the content of each attribute becomes clearer

In the PR description, you mention that you went this way because env.get_parents() doesn't make sense for some envs like ctorus. Could this be decided in Batch by looking at the env instead of the loss? Like, if the env has a get_parents() implemented, then you use it and collect the possible parents of each state. If it doesn't, then you don't (and, therefore, that information is not made available to the GFlowNet so it will need to pick a loss other than FM for such envs).

Reply:
I see the reason behind your point, but I would prefer to inform batch about the loss and collect only the necessary information. apart from the issue with `get_parrents`, it is also more efficient to collect only the things which are needed for computing the loss, especially for long trajectories 

As for changing the semantics of the attributes names, maybe we can resolve it by having two different attributes: `all_possible_parents` and `parents` ?

Reply:
Would it make sense if the Batch class had arguments like "collect_X", "collect_Y", ... for the various optional things that could be collected but won't always be required? Like for the possible parents, for instance.
Then, instead of providing the loss to the Batch class, the GFlowNet class can instead simply tell the Batch class what should be collected.
I think that this achieves your goal of efficiency but it also makes Batch agnostic to the loss. For instance, if we ever implement a new loss or new performance metrics, we only need to make the change in GFlowNet and GFlowNet will tell Batch what needs to be collected to compute this new loss/metric.

For the semantics, yes, I think that the way you are proposing would much clearer.

Reply:
Would it make sense if the Batch class had arguments like "collect_X", "collect_Y", ... for the various optional things that could be collected but won't always be required? Like for the possible parents, for instance.
Then, instead of providing the loss to the Batch class, the GFlowNet class can instead simply tell the Batch class what should be collected.
I think that this achieves your goal of efficiency but it also makes Batch agnostic to the loss. For instance, if we ever implement a new loss or new performance metrics, we only need to make the change in GFlowNet and GFlowNet will tell Batch what needs to be collected to compute this new loss/metric.

For the semantics, yes, I think that the way you are proposing would much clearer.

Reply:
The new names look good.
For the additional arguments, that's fair. I mean, the current PR is still a good improvement over the current state and there's nothing preventing a revisit in the future.

Reply:
I think we should eventually make the Batch independent of the loss, as Pierre-Luc argued, but I would do this later. No needed for this PR IMO.